dashboard/config/initializers/refresh_activemodel_cached_attributes.rb

```diff
@@ -0,0 +1,72 @@
+module Cdo
+  # Rails caches information about valid attributes based on the state of the
+  # database at initialization time, which means if a new migration has run on
+  # the database since then, that cache might be operating off of old
+  # information. Something like `Model.new(Model.last.attributes)` will result
+  # in `Model.new` using stale validation logic against fresh attributes from
+  # `Model.last`, which may not work.
+  #
+  # To resolve, if we get an UnknownAttributeError, we first check to see if
+  # there have been any new migrations executed against the database since this
+  # instance of Rails was initialized. If there have been, we attempt to
+  # refresh some caches and try again before treating this as an actual error.
+  #
+  # Note that this solution was crafted for Ruby on Rails v6.1.7.7, and is
+  # unfortunately reliant on some implementation details which may change in
+  # later versions.
```

Comment:
Is this something we should considering upstreaming? I'll say "today's fix that touches internals for a small gain is tomorrow's major hassle or breakage", but I'm biased by recent version upgrading. Overall one of the lessons I'm learning at code.org is that much of the hard work done by people to make things "perfect" in the past that caused us to deviate from Rails default improved things for a while, and then became a downside in unexpected ways.

Reply:
I think that's definitely a valid concern!

One thing that makes me feel a bit more comfortable about this change in particular is that it's not adding any new functionality that we might come to depend on, it's just providing us protection against a particular footgun that can arise in a particular situation. If we do need to disable this for any reason in the future, we should be able to do so without breaking any existing code; we'd just then again need to be careful about migrations that rename or add columns.

Alternatively, we could try to figure out an automated way to defend against that dangerous use of `.attributes` and warn developers of the implications it has for migrations. That feels even more difficult to maintain, though.

Getting this merged upstream definitely sounds like the best path forward! Frustratingly, I couldn't find any evidence online of other people running into this particular problem, even though I don't think it's being caused by anything specific to our setup. Opening up a pull request and soliciting some feedback seems like a good way to find some of that evidence if it exists, though!

Reply:
I believe we could address most of these concerns by using the Rails [Active Record Store](https://api.rubyonrails.org/classes/ActiveRecord/Store.html) feature.
```ruby
class User < ApplicationRecord
  store :properties, accessors: [...], coder: JSON
end
```

Reply:
I definitely think it would make sense for us to revisit our `SharedProperties` implementation with a view to possibly transitioning it to an Active Record Store, but that's a separate issue. This is addressing some details about ActiveModel's underlying implementation of attributes, which Active Record is built on top of.